### ISSUE TITLE
[blog post] JS Event Loop

### ISSUE THREAD

**Issue by dev-cprice** [dev-cprice/codyaprice.com#87]
* why it matters
  * single threaded runtime model
  * code runs to completion and then terminates
* tie into FDM code
* single threading
* don’t block it

![image](https://user-images.githubusercontent.com/34115417/72672205-6f5bbd80-3a1c-11ea-8aa2-8ac318232e74.png)

## FDM Code (ish)

```c
// 32 flags used to determine when to run an event handler,
// initially zeroed out
// flags will be set via hardware interrupts
uint32_t eventFlags = 0;

// list of 32 event handlers, as function pointers (references)
void* eventHandlers[32] = {
  onNetworkChange,
  onNetworkRequest,
  onTemperatureChange,
  onVoltageDrop,
  // 28 other event handlers
};

// check all event flags, if set run the event handler
void handleEvents() {
  for (uint8_t i = 0; i < 32; ++i) {
    // get the flag at this position
    uint8_t flag = eventFlags >> i & 0x01;

    // if set, run the handler #i
    if (flag) {
      eventHandlers[i]();
    }
  }

  // clear all the event bits
  eventFlags = 0;
}

int main() {
  // our event loop
  while (1) {
    handleEvents();
    sleep(500);
  }
}
```

## Jake Archibald Event Loop Talk Notes
* `setTimeout(callback, ms)`
	1. Run the following steps in parallel (off the main thread):
		1. Wait /ms/ milliseconds
		2. Queue a task to run the following steps:
			1. Invoke /callback/
* task queues
* the render steps
	* style calc
	* layout (render tree, positioning)
	* painting
* Example: button.onClick = () => while (true);
	* blocking in task queue items still blocks the main thread!
* tasks
	* must run to completion before next render
* setTimeout loop, not render blocking
```js
function loop() {
  setTimeout(loop, 0);
}
loop();
```
* `requestAnimationFrame`, happen as part of the render steps
* if a tab is in the background, the render step never happens
* `setTimeout` can end up with drift, nothing in one frame, twice the work in others… can move the render steps around if they run too long
* `raf` can be used to batch work within a single frame, save a lot of duplicate work
* micro tasks
	* run when the stack is empty
		* anywhere JS can run
	* promises use them
```js
// this blocks the main thread
function loop() {
  Promise.resolve().then(loop);
}
loop();
```
	* doesn’t yield to rendering
* Task Queue
	* one item only at a time, new things go to the end of the queue
* Animation Callbacks Queue
	* run to completion, except those queued during this loop: deferred to next frame
* Microtasks
	* run to completion, including those queued during this loop
	* if you add micro tasks as fast as you complete them, this runs infinitely
* clicking via JS has side effects/consequences
```js
button.addEventListener("click", () => {
  Promise.resolve().then(() => console.log("microtask 1"));
  console.log("listener 1");
});

button.addEventListener("click", () => {
  Promise.resolve().then(() => console.log("microtask 2"));
  console.log("listener 2");
});

// click on the button via the mouse

// #=> listener 1
// #=> microtask 1
// #=> listener 2
// #=> microtask 2
```
```js
button.addEventListener("click", () => {
  Promise.resolve().then(() => console.log("microtask 1"));
  console.log("listener 1");
});

button.addEventListener("click", () => {
  Promise.resolve().then(() => console.log("microtask 2"));
  console.log("listener 2");
});

button.click(); // js stack ain't empty till this returns, doesn't return until all listeners are done

// #=> listener 1
// #=> listener 2
// #=> microtask 1
// #=> microtask 2
```

## Philip Roberts Event Loop Talk Notes

JS, what are you?
	* a single-threaded, non-blocking asynchronous concurrent language
	* I have a call stack, an event loop, a callback queue, some other apis and stuff
* v8 just has a call stack and heap, nothing else
	* setTimeout, DOM, etc aren’t a part of v8, but in the /browser/
	* WebAPIs
		* DOM
		* ajax
		* setTimeout
* the call stack
	* one thread == one call stack == one thing at a time
	* records where in the function we are
```js
function multiply(a, b) {
  return a * b;
}

function square(n) {
  return multiply(n, n);
}

function printsquare(n) {
  var squared = square(n);
  console.log(squared);
}

printSquare(4);
```
* blocking
	* what happens when things are slow?
	* image processing, 1 billion iteration loop, network requests are slow
	* the solution? Asynchronous callbacks
* Concurrency and the event loop. One thing at a time, except not really.
